pkg/util/machine/network_linux.go

```diff
@@ -0,0 +1,272 @@
+//go:build linux
+// +build linux
+
+/*
+Copyright 2022 The Katalyst Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machine
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"syscall"
+
+	"github.com/vishvananda/netns"
+
+	"github.com/kubewharf/katalyst-core/pkg/config/agent/global"
+	"github.com/kubewharf/katalyst-core/pkg/util/general"
+)
+
+const (
+	sysFSDirNormal   = "/sys"
+	sysFSDirNetNSTmp = "/tmp/net_ns_sysfs"
+)
+
+const (
+	nicPathNameDeviceFormatPCI = "devices/pci"
+	nicPathNAMEBaseDir         = "class/net"
+)
+
+const (
+	netFileNameSpeed    = "speed"
+	netFileNameNUMANode = "device/numa_node"
+	netFileNameEnable   = "device/enable"
+	netEnable           = 1
+)
+
+// GetExtraNetworkInfo get network info from /sys/class/net and system function net.Interfaces.
+// if multiple network namespace is enabled, we should exec into all namespaces and parse nics for them.
+func GetExtraNetworkInfo(conf *global.MachineInfoConfiguration) (*ExtraNetworkInfo, error) {
+	networkInfo := &ExtraNetworkInfo{}
+
+	nsList := []string{DefaultNICNamespace}
+	if conf.NetMultipleNS {
+		if conf.NetNSDirAbsPath == "" {
+			return nil, fmt.Errorf("GetNetworkInterfaces got nil netNSDirAbsPath")
+		}
+
+		if dirs, err := ioutil.ReadDir(conf.NetNSDirAbsPath); err != nil {
+			return nil, err
+		} else {
+			for _, dir := range dirs {
+				if !dir.IsDir() {
+					nsList = append(nsList, dir.Name())
+				}
+			}
+		}
+
+		runtime.LockOSThread()
+		defer runtime.UnlockOSThread()
+	}
+
+	for _, ns := range nsList {
+		nicsInNs, err := getNSNetworkHardwareTopology(ns, conf.NetNSDirAbsPath)
+		if err != nil {
+			// if multiple ns is disabled, we should block agent start
+			// todo: to discuss if block agent start when any ns is parsed failed
+			if !conf.NetMultipleNS {
+				return nil, fmt.Errorf("failed to get network topology: %v", err)
+			}
+
+			general.Errorf("get network topology for ns %s failed: %v", ns, err)
+			continue
+		}
+
+		networkInfo.Interface = append(networkInfo.Interface, nicsInNs...)
```

Comment:
1. Interface -> Interfaces may be better?  since there may be multi-nics in the machine 
2. why not use []*machine.InterfaceInfo type